phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java

```diff
@@ -1087,116 +1087,140 @@ private static PTable deserializeTable(byte[] b) {
+                                        del.setAttribute(BaseScannerRegionObserver.REPLAY_WRITES,
+                                                BaseScannerRegionObserver.REPLAY_INDEX_REBUILD_WRITES);
+                                        del.setAttribute(BaseScannerRegionObserver.CLIENT_VERSION, clientVersionBytes);
+                                        mutations.add(del);
+                                        // Since we're replaying existing mutations, it makes no sense to write them to the wal
+                                        del.setDurability(Durability.SKIP_WAL);
+                                    }
+                                    del.addDeleteMarker(cell);
                                 }
-                                del.addDeleteMarker(cell);
                             }
+                            if (ServerUtil.readyToCommit(mutations.size(), mutations.byteSize(), maxBatchSize, maxBatchSizeBytes)) {
+                                checkForRegionClosing();
+                                commitBatchWithRetries(region, mutations, blockingMemstoreSize);
+                                uuidValue = ServerCacheClient.generateId();
+                                mutations.clear();
+                            }
+                            rowCount++;
                         }
-                        if (ServerUtil.readyToCommit(mutations.size(), mutations.byteSize(), maxBatchSize, maxBatchSizeBytes)) {
-                            checkForRegionClosing();
-                            commitBatchWithRetries(region, mutations, blockingMemstoreSize);
-                            uuidValue = ServerCacheClient.generateId();
-                            mutations.clear();
-                        }
-                        rowCount++;
+
+                    } while (hasMore && rowCount < pageSizeInRows);
+                    if (!mutations.isEmpty()) {
+                        checkForRegionClosing();
+                        commitBatchWithRetries(region, mutations, blockingMemstoreSize);
                     }
-                    
-                } while (hasMore);
-                if (!mutations.isEmpty()) {
-                    checkForRegionClosing();
-                    commitBatchWithRetries(region, mutations, blockingMemstoreSize);
+                }
+            } catch (IOException e) {
+                hasMore = false;
+                LOGGER.error("IOException during rebuilding: " + Throwables.getStackTraceAsString(e));
+                throw e;
+            } finally {
+                if (!hasMore) {
+                    region.closeRegionOperation();
                 }
             }
-        } catch (IOException e) {
-            LOGGER.error("IOException during rebuilding: " + Throwables.getStackTraceAsString(e));
-            throw e;
-        } finally {
-            region.closeRegionOperation();
+            byte[] rowCountBytes = PLong.INSTANCE.toBytes(Long.valueOf(rowCount));
```

Comment:
Does rowCount here mean "the total count of rows successfully committed by commitBatchWithRetries"? If that's the case, we might have inconsistency issue -- suppose above loop called commitBatchWithRetries multiple times but the last one failed (thrown IOException) then rowCount here counted those uncommitted rows -- is this expected?

Reply:
The row count is used to show the progress for the MR job. It is not really for correctness. However, if there is an exception, then the mapper task will get an exception and no row count will be returned to the mapper. The mapper task will start from the beginning to rebuild those rows again. When that happens, I expect the row count will be wrong because this mapper task will rebuild and double count for the rows that are rebuilt again. I will think about improving this, maybe not in this JIRA.